frontend/src/lib/logic/apiStatusLogic.ts

```diff
@@ -0,0 +1,53 @@
+import { actions, kea, listeners, path, reducers } from 'kea'
+import api from 'lib/api'
+
+import type { apiStatusLogicType } from './apiStatusLogicType'
+
+export const apiStatusLogic = kea<apiStatusLogicType>([
+    path(['lib', 'apiStatusLogic']),
+    actions({
+        onApiResponse: (response?: Response, error?: any) => ({ response, error }),
+        setInternetConnectionIssue: (issue: boolean) => ({ issue }),
+    }),
+    reducers({
+        internetConnectionIssue: [
+            false,
+            {
+                setInternetConnectionIssue: (_, { issue }) => issue,
+            },
+        ],
+    }),
+    listeners(({ cache, actions, values }) => ({
+        onApiResponse: async ({ response, error }, breakpoint) => {
+            if (error || !response?.status) {
+                await breakpoint(50)
+                // Likely CORS headers errors (i.e. request failing without reaching Django))
+                if (error?.message === 'Failed to fetch') {
+                    actions.setInternetConnectionIssue(true)
+                }
+            }
+
+            if (response?.ok && values.internetConnectionIssue) {
+                actions.setInternetConnectionIssue(false)
+            }
+
+            if (response?.status === 401) {
+                // api.ts calls this if we see a 401
+                const now = Date.now()
+
+                // We don't want to check too often (multiple api calls might fail at once)
+                if (now - 10000 > (cache.lastUnauthorizedCheck ?? 0)) {
+                    cache.lastUnauthorizedCheck = Date.now()
+
+                    await breakpoint(100)
+
+                    await api.get('api/users/@me/').catch((error: any) => {
+                        if (error.status === 401) {
+                            window.location.href = '/logout'
+                        }
+                    })
```

Comment:
Weirdly I've not been able to get this to work. Can we have a Cypress test for this logic? There are a lot of building blocks here, so I think E2E would be useful, and at the same time the intended behavior is straightforward so should be easy to test (`fetch('/logout')`, then any click should redirect to login)

Reply:
It works perfectly for me. Two tabs logged in. Log out of 1 and then click something in the other and it will trigger this flow. It's hard to test in a mocked way as there is other routing that gets in the way.

I'm not against a cypress test - I just have no idea how one would simulate that!

Reply:
Looks like I just had some unknown local problem, added a test (take a look to verify) and now the logic works perfectly (confirmed the test only passes on this branch and fails on master)